### PR TITLE
Highlight mention when cursor is moved inside of it.

### PR DESCRIPTION
**Pull request by aniyati**
- When cursor is placed within mention, highlight that mention.

**Testing**
- Tested mention highlighting manually with `enableSimpleRefactor` enabled state.

**Attachments**
|  1  |  2 (Tested with Normal Text) |
| --- | --- |
| gif1 | gif2 |
| HighlightedMention_MentionEdited  | |

### CODE REVIEW

File: Hakawai/Mentions/HKWMentionsPlugin.m [hunk 1]
```diff
@@ -1677,6 +1677,15 @@ - (void)textViewDidChangeSelectionV2:(UITextView *)textView {
     if (range.length > 0) {
         return;
     }
+
+    // When cursor is placed within mention, highlight that mention.
+    const NSUInteger cursorLocation = range.location;
+    [self.mentions enumerateObjectsUsingBlock:^(HKWMentionsAttribute *mention, NSUInteger idx __unused, BOOL *stop __unused) {
+        const NSUInteger mentionStartLocation = mention.range.location;
+        const BOOL isCursorInsideMention = (cursorLocation > mentionStartLocation) && (cursorLocation < mentionStartLocation + mention.range.length);
+        [self toggleMentionsFormattingAtRange:mention.range selected:isCursorInsideMention];
```
Comment (bkoatz) [inline Hakawai/Mentions/HKWMentionsPlugin.m:1686]: ok so after looking at this, maybe using currentlySelectedMentionRange is good.

So in this function, you can do the following

NSRange range;
[attribute:HKWMentionsAttribute atIndex:range.location effectiveRange:&range];

if range != self.currentlySelectionMentionRange
   [self toggleMentionsFormattingAtRange:currentlySelectionMentionRange selected:NO];
   [self toggleMentionsFormattingAtRange:range selected:YES]; 
   
self.currentlySelectionMentionRange = range

with appropriate guarding

File: Hakawai/Mentions/HKWMentionsPlugin.m [hunk 2]
```diff
@@ -1677,6 +1677,15 @@ - (void)textViewDidChangeSelectionV2:(UITextView *)textView {
     if (range.length > 0) {
         return;
     }
+
+    // When cursor is placed within mention, highlight that mention.
```
Comment (bkoatz) [inline Hakawai/Mentions/HKWMentionsPlugin.m:1681]: also could you put this in a separate function?

File: Hakawai/Mentions/HKWMentionsPlugin.m (hunk 1 above)
Comment (aniyati) [reply Hakawai/Mentions/HKWMentionsPlugin.m]: Thanks @bkoatz  for reviewing PR.
 
We will be using `attribute(at:effectiveRange:)` as described in https://github.com/linkedin/Hakawai/pull/170#discussion_r461786851

- On selecting mention, we want to highlight that mention and de-select previous selected mention.
- `currentlySelectedMentionRange` holds current mention range only and it's location always start with next location (i.e `+1` ) than actual mention location.
- So, we have to maintain previous selected mention to de-select it or de-select all mentions expect one that is going to be selected.
- So we think that usage of `currentlySelectedMentionRange`  won't be suitable in this case.

Please let us know if we are missing something.

File: Hakawai/Mentions/HKWMentionsPlugin.m (hunk 2 above)
Comment (aniyati) [reply Hakawai/Mentions/HKWMentionsPlugin.m]: Moved code in separate function.

File: Hakawai/Mentions/HKWMentionsPlugin.m (hunk 1 above)
Comment (bkoatz) [reply Hakawai/Mentions/HKWMentionsPlugin.m]: I think just whatever range you get from attribute(at:...) you can save for future reference so you can unselect it later. you can save it in a variable currentlySelectedMentionRangeV2, which I referenced in slack.

Comment (aniyati) [reply Hakawai/Mentions/HKWMentionsPlugin.m]: We have tried to implement `attribute(at:effectiveRange:)` but this method has issue with normal text.
`attribute(at:effectiveRange:)` return mention range for normal text as well.
So we have used similar method (i.e `enumerateAttributesInRange(:options:usingBlock)` which also iterates over attributes) and not considers normal text as mention text.

Also introduced `currentlySelectedMentionRangeV2` property.

File: Hakawai/Mentions/HKWMentionsPlugin.m [hunk 3]
```diff
@@ -846,6 +853,52 @@ - (void)toggleAutocorrectAsRequiredForRange:(NSRange)range {
...
 
+- (void)highlightMentionIfNeededForCursorLocation:(NSUInteger)cursorLocation {
+    // When cursor is placed within mention, highlight that mention.
+    __strong __auto_type parentTextView = self.parentTextView;
+    const NSRange textFullRange = HKW_FULL_RANGE(parentTextView.attributedText);
+
+    // When whitespace is inserted at end of text, it is not included in text range.
+    // So check if cursor location is inside text range.
+    if (NSLocationInRange(cursorLocation, textFullRange)) {
+        const NSRange currentMentionRange = [HKWMentionsPlugin mentionRangeAtLocation:cursorLocation attributedText:parentTextView.attributedText];
```
Comment (bkoatz) [inline Hakawai/Mentions/HKWMentionsPlugin.m:864]: https://developer.apple.com/documentation/foundation/nsattributedstring/1408174-attribute

I think you can use this. HKWMentionsAttribute is the attribute. And then you don't need to have the extra function! And then you can do stuff based on the equal range == currentlySelectedMentionRangeV2?

Comment (aniyati) [reply Hakawai/Mentions/HKWMentionsPlugin.m:864]: We have tried to implement `attribute(at:effectiveRange:)` but this method has issue when we move over normal text.
`attribute(at:effectiveRange:)` return mention range for normal text as well.
So we have used similar method (i.e `enumerateAttributesInRange(:options:usingBlock)` which also iterates over attributes) and not considers normal text as mention text.

**Comment by bkoatz** [comment]
moved logic into here https://github.com/linkedin/Hakawai/pull/178/files thank u for this!

**Pull request closed without merging**